core/DataAccess/LogQueryBuilder/JoinGenerator.php

```diff
@@ -0,0 +1,264 @@
+
+                if ($this->tables->hasJoinedTable($tableNameToJoin)) {
+                    $this->generateNonVisitJoins($table, $tableNameToJoin, $index);
+                }
+            }
+        }
+    }
+
+    /**
+     * Generate the join sql based on the needed tables
+     * @throws Exception if tables can't be joined
+     * @return array
+     */
+    public function generate()
+    {
+        /** @var LogTable[] $availableLogTables */
+        $availableLogTables = array();
+
+        $this->tables->sort(array($this, 'sortTablesForJoin'));
+
+        foreach ($this->tables as $i => $table) {
+            if (is_array($table)) {
+
+                // join condition provided
+                $alias = isset($table['tableAlias']) ? $table['tableAlias'] : $table['table'];
+                $this->joinString .= " LEFT JOIN " . Common::prefixTable($table['table']) . " AS " . $alias
+                                   . " ON " . $table['joinOn'];
+                continue;
+            }
+
+            $tableSql = Common::prefixTable($table) . " AS $table";
+
+            $logTable = $this->tables->getLogTable($table);
+
+            if ($i == 0) {
+                // first table
+                $this->joinString .= $tableSql;
+            } else {
+
+                $join = $this->findJoinCriteriasForTables($logTable, $availableLogTables);
+
+                if ($join === null) {
+                    $availableLogTables[$table] = $logTable;
+                    continue;
+                }
+
+                // the join sql the default way
+                $this->joinString .= " LEFT JOIN $tableSql ON " . $join;
+            }
+
+            $availableLogTables[$table] = $logTable;
+        }
+    }
+
+    public function getJoinString()
+    {
+        return $this->joinString;
+    }
+
+    public function shouldJoinWithSelect()
+    {
+        return $this->joinWithSubSelect;
+    }
+
+    /**
+     * @param LogTable $logTable
+     * @param LogTable[] $availableLogTables
+     * @return string|null   returns null in case the table is already joined, or the join string if the table needs
+     *                       to be joined
+     * @throws Exception if table cannot be joined for segmentation
+     */
+    protected function findJoinCriteriasForTables(LogTable $logTable, $availableLogTables)
+    {
+        $join = null;
+        $alternativeJoin = null;
+        $table = $logTable->getName();
+
+        foreach ($availableLogTables as $availableLogTable) {
+            if ($logTable->getColumnToJoinOnIdVisit() && $availableLogTable->getColumnToJoinOnIdVisit()) {
+
+                $join = sprintf("%s.%s = %s.%s", $table, $logTable->getColumnToJoinOnIdVisit(),
+                                                 $availableLogTable->getName(), $availableLogTable->getColumnToJoinOnIdVisit());
+                $alternativeJoin = sprintf("%s.%s = %s.%s", $availableLogTable->getName(), $availableLogTable->getColumnToJoinOnIdVisit(),
```

Comment:
This could eg fix some issues when creating an archiver. If they "ON" clause was written the other way around in the Archiver, it would have resulted in a bug as we would not have detected that the "ON" clause already exists